cli/src/config.rs

```diff
@@ -17,33 +18,79 @@ To learn more, checkout https://apollo.dev/cli/telemetry\n";
 #[derive(Debug, Serialize, Deserialize)]
 pub struct CliConfig {
     pub machine_id: String,
+    // This can be used for a service api key, or a personal api key.
+    // The CLI doesn't differentiate at the moment.
+    pub api_key: Option<String>,
+}
+
+fn save(path: &PathBuf, cli_config: &CliConfig) -> Result<(), Box<dyn Error + 'static>> {
+    let toml = toml::to_string(cli_config).unwrap();
+
+    fs::create_dir_all(&path.parent().unwrap())?;
+    debug!("Wrote cli config to path {}", path.to_str().unwrap());
+    fs::write(&path, toml).map_err(From::from)
+}
+
+fn load(path: &PathBuf) -> Result<CliConfig, Box<dyn Error + 'static>> {
+    let mut s = Config::new();
+
+    if !path.exists() {
+        CliConfig::create(path)?;
+    }
+
+    s.merge(config::File::with_name(path.to_str().unwrap()))?;
+
+    s.merge(Environment::with_prefix("APOLLO_").separator("__"))
+        .unwrap();
+
+    debug!("Loaded cli config from path {}", path.to_str().unwrap());
+    s.try_into().map_err(From::from)
 }
 
 impl CliConfig {
-    pub fn load() -> Result<Self, Box<dyn Error + 'static>> {
-        let mut s = Config::new();
+    pub fn save(path: &PathBuf, cli_config: &CliConfig) -> Fallible<()> {
+        save(path, cli_config).map_err(|e| {
+            ApolloError::from(ErrorDetails::CliConfigWriteError {
+                msg: e.to_string(),
+                path: path.to_str().unwrap().to_string(),
+            })
+        })
+    }
 
-        let config_path = apollo_config().unwrap();
+    pub fn create(path: &PathBuf) -> Result<Self, Box<dyn Error + 'static>> {
+        let config = CliConfig {
+            machine_id: Uuid::new_v4().to_string(),
+            api_key: None,
+        };
 
-        if config_path.exists() {
-            s.merge(config::File::with_name(config_path.to_str().unwrap()))?;
-        } else {
-            // create new file with uuid
-            let machine_id = Uuid::new_v4().to_string();
-            s.set("machine_id", machine_id)?;
+        debug!(
+            "New cli config: {}",
+            serde_json::to_string(&config).unwrap()
+        );
 
-            let generated: Result<CliConfig, ConfigError> = s.clone().try_into();
+        CliConfig::save(path, &config)?;
 
-            let toml = toml::to_string(&generated.unwrap()).unwrap();
+        // log initial telemetry warning
+        info!("{}", POST_INSTALL_MESSAGE);
+        Ok(config)
+    }
 
-            fs::create_dir_all(&config_path.parent().unwrap())?;
-            fs::write(&config_path, toml)?;
+    pub fn load(path: &PathBuf) -> Fallible<Self> {
+        load(path).map_err(|e| {
```

Comment:
this closure is repeated, consider extracting to a function for DRYer code?

Reply:
I think we can actually make this much more concise in a number of ways, depending on whether we feel that the separate details for writing and reading are valuable? ie, if we could refactor the `CliConfigWriteError` and `CliConfigLoadError` into `CliConfigError` where we say something like: 

```
Configuration error for file {}: {}
```

We could also have an ENUM of "actions", but I'm not sure how valuable it really is specifying whether it's a read/write issue etc. (especially because I suspect this will frequently be a permissions or not found issue)

Reply:
Hm, I don't know how to answer that; Maybe that's worthwhile but in the spirit of keeping changes small, I just mean that this can be something like
```rust
load(path).map_err(|e| read_error(e, path))
```

Reply:
I think I'm confused now, wouldn't we then need a `read_error` and a `write_error`? Since the two closures which do the error mapping are building different errors themselves (one a `CliConfigWriteError` the other a `CliConfigReadError`) we'd need have the shared code be able to build one or the other dynamically no?

Reply:
Oh sorry, maybe I didn't notice they're different error types!